### PR TITLE
Fix loss_bbox in fovea_head.py when num_pos=0

### PR DESCRIPTION
**Pull request by anurag1paul**
During multi-GPU training, when num_pos == 0 on line 216 in fovea_head.py, it causes loss_bbox to be a tensor vector of shape [1] while when num_pos > 0, it causes loss_bbox to be a scalar of shape [], causing the gather function to fail. 

This pull request fixes this bug by unsqueezing loss_bbox when num_pos > 0.

### CODE REVIEW

File: .isort.cfg [hunk 1]
```diff
@@ -3,6 +3,6 @@ line_length = 79
 multi_line_output = 0
 known_standard_library = setuptools
 known_first_party = mmdet
-known_third_party = asynctest,cityscapesscripts,cv2,matplotlib,mmcv,numpy,onnx,pycocotools,robustness_eval,roi_align,roi_pool,seaborn,six,terminaltables,torch,torchvision
+known_third_party = PIL,asynctest,cityscapesscripts,cv2,matplotlib,mmcv,numpy,onnx,pycocotools,robustness_eval,roi_align,roi_pool,seaborn,six,terminaltables,torch,torchvision
```
Comment (ZwwWayne) [inline .isort.cfg:6]: It seems that we usually do not use PIL in our codebase, may remove this.

Comment (anurag1paul) [reply .isort.cfg:6]: Should we remove it from runtime requirements as well? 
https://github.com/open-mmlab/mmdetection/blob/c40cafb1e1347645c1289634a40cc26b7b58163c/requirements/runtime.txt#L5

I committed using an environment generated using requirements files and pre-commit check forced me to add this to commit before it accepted the commit.

Comment (ZwwWayne) [reply .isort.cfg:6]: No, we cannot remove it from the requirements because of the potential incompatibility of torchvision and PILLOW versions.

Comment (ZwwWayne) [reply .isort.cfg:6]: But the mmdet code indeed does not use PIL, it is wierd.

Comment (anurag1paul) [reply .isort.cfg]: I have removed PIL from .isort.cfg.

File: mmdet/models/anchor_heads/fovea_head.py [hunk 2]
```diff
@@ -223,6 +223,7 @@ def loss(self,
                 pos_bbox_targets,
                 pos_weights,
                 avg_factor=num_pos)
+            loss_bbox = loss_bbox.unsqueeze(0)
```
Comment (ZwwWayne) [inline mmdet/models/anchor_heads/fovea_head.py:226]: Unsqueezing the `loss_bbox` when `num_pos>0` is a little bit risky since it is more commonly used. It seems that other detectors' losses also do not have such a pattern.
So maybe it is better and safer to squeeze the loss_bbox when `num_pos==0`?

Comment (anurag1paul) [reply mmdet/models/anchor_heads/fovea_head.py:226]: Squeezing when num_pos==0 would make loss a scalar, which sometimes causes pytorch to throw a warning "UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector."
Otherwise, I do not see why the author chose:
`loss_bbox = torch.tensor([0],
                                     dtype=flatten_bbox_preds.dtype,
                                     device=flatten_bbox_preds.device)`

and not
`loss_bbox = torch.tensor(0,
                                     dtype=flatten_bbox_preds.dtype,
                                     device=flatten_bbox_preds.device)`

Comment (ZwwWayne) [reply mmdet/models/anchor_heads/fovea_head.py:226]: However, in the codebases, the final losses are always a scalar. 
So we suggest changing the case`num_pos==0` and use either
```python
loss_bbox = torch.tensor([0], dtype=flatten_bbox_preds.dtype, device=flatten_bbox_preds.device).sum()
```
or 
```python
loss_bbox = torch.tensor(0, dtype=flatten_bbox_preds.dtype, device=flatten_bbox_preds.device)
```
Does this work for you?

Comment (anurag1paul) [reply mmdet/models/anchor_heads/fovea_head.py:226]: Sure, I will do that.

Comment (ZwwWayne) [reply mmdet/models/anchor_heads/fovea_head.py]: Thanks.

**Comment by ZwwWayne** [comment]
Hi @anurag1paul ,
Thanks for your kind contribution. I notice that your latest commit changes the code from `loss_bbox = torch.tensor(0)` to `loss_bbox = torch.tensor([0]).sum()`, do you change it for any specific reason? For example, do you meet warning or bug when using `loss_bbox = torch.tensor(0)`?

**Comment by anurag1paul** [comment]
With 'loss_bbox = torch.tensor(0)`, the build was Travis PR build was erroring out, so I tried your second suggestion and it was able to pass the Travis CI checks.

**Comment by ZwwWayne** [comment]
> With 'loss_bbox = torch.tensor(0)`, the Travis PR build was erroring out, so I tried your second suggestion and it was able to pass the Travis CI checks.

OK, thank you.

**Comment by anurag1paul** [comment]
> > With 'loss_bbox = torch.tensor(0)`, the Travis PR build was erroring out, so I tried your second suggestion and it was able to pass the Travis CI checks.
> 
> OK, thank you.
> But I check the CI, it seems that 'loss_bbox = torch.tensor(0)` cannot pass the CI only because of the format cannot pass yapf and it should works after you reformat the code?

Yeah, I checked. It does seem a yapf issue. I am having a lot of trouble committing as pre-commit keeps adding PIL to .isort.cfg. Can you help me in skipping "seed isort known_third_party"?

**Comment by ZwwWayne** [comment]
> > > With 'loss_bbox = torch.tensor(0)`, the Travis PR build was erroring out, so I tried your second suggestion and it was able to pass the Travis CI checks.
> > 
> > 
> > OK, thank you.
> > But I check the CI, it seems that 'loss_bbox = torch.tensor(0)` cannot pass the CI only because of the format cannot pass yapf and it should works after you reformat the code?
> 
> Yeah, I checked. It does seem a yapf issue. I am having a lot of trouble committing as pre-commit keeps adding PIL to .isort.cfg. Can you help me in skipping "seed isort known_third_party"?

Add `-n` in your command should work. It seems that `xml_dataset` indeed use PIL, so you can also allow the pre-commit to add it. I missed that in the format comments.

**Comment by anurag1paul** [comment]
> > > > With 'loss_bbox = torch.tensor(0)`, the Travis PR build was erroring out, so I tried your second suggestion and it was able to pass the Travis CI checks.
> > > 
> > > 
> > > OK, thank you.
> > > But I check the CI, it seems that 'loss_bbox = torch.tensor(0)` cannot pass the CI only because of the format cannot pass yapf and it should works after you reformat the code?
> > 
> > 
> > Yeah, I checked. It does seem a yapf issue. I am having a lot of trouble committing as pre-commit keeps adding PIL to .isort.cfg. Can you help me in skipping "seed isort known_third_party"?
> 
> Add `-n` in your command should work. It seems that `xml_dataset` indeed use PIL, so you can also allow the pre-commit to add it. I missed that in the format comments.

Thank you. I have added PIL to .isort-cfg to make life easier with pre-commit.

**Comment by ZwwWayne** [comment]
Thanks very much.

**Comment by openmmlab-v** [comment]
Hi @anurag1paul ！First of all, we want to express our gratitude for your significant PR in the mmdetection project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you are Chinese or have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
